x-pack/plugins/cases/public/components/actions/tags/edit_tags_selectable.tsx

```diff
@@ -0,0 +1,398 @@
+
+      for (const tag of action.payload) {
+        selectedTags[tag] = { tagState: TagState.CHECKED, dirty: true, icon: ICONS.CHECKED };
+      }
+
+      return { ...state, tags: { ...state.tags, ...selectedTags } };
+
+    case Actions.UNCHECK_TAG:
+      const unselectedTags: State['tags'] = {};
+
+      for (const tag of action.payload) {
+        unselectedTags[tag] = { tagState: TagState.UNCHECKED, dirty: true, icon: ICONS.UNCHECKED };
+      }
+
+      return { ...state, tags: { ...state.tags, ...unselectedTags } };
+
+    default:
+      assertNever(action);
+  }
+};
+
+const getInitialTagsState = ({
+  tags,
+  selectedCases,
+}: {
+  tags: string[];
+  selectedCases: Case[];
+}): State => {
+  const tagCounterMap = createTagsCounterMapping(selectedCases);
+  const totalCases = selectedCases.length;
+  const tagsRecord: State['tags'] = {};
+  const state = { tags: tagsRecord, tagCounterMap };
+
+  for (const tag of tags) {
+    const tagCounter = tagCounterMap.get(tag) ?? 0;
+    const isCheckedTag = tagCounter === totalCases;
+    const isPartialTag = tagCounter < totalCases && tagCounter !== 0;
+    const tagState = isCheckedTag
+      ? TagState.CHECKED
+      : isPartialTag
+      ? TagState.PARTIAL
+      : TagState.UNCHECKED;
+
+    const icon = getSelectionIcon(tagState);
+
+    tagsRecord[tag] = { tagState, dirty: isCheckedTag, icon };
+  }
+
+  return state;
+};
+
+const createTagsCounterMapping = (selectedCases: Case[]) => {
+  const counterMap = new Map<string, number>();
+
+  for (const theCase of selectedCases) {
+    const caseTags = theCase.tags;
+
+    for (const tag of caseTags) {
+      counterMap.set(tag, (counterMap.get(tag) ?? 0) + 1);
+    }
+  }
+
+  return counterMap;
+};
+
+const stateToOptions = (tagsState: State['tags']): TagSelectableOption[] => {
+  const tags = Object.keys(tagsState);
+
+  return tags.map((tag): EuiSelectableOption => {
+    return {
+      key: tag,
+      label: tag,
+      ...(tagsState[tag].tagState === TagState.CHECKED ? { checked: 'on' } : {}),
+      'data-test-subj': `cases-actions-tags-edit-selectable-tag-${tag}`,
+      data: { tagIcon: tagsState[tag].icon },
+    };
+  }) as TagSelectableOption[];
+};
+
+const getSelectionIcon = (tagState: TagState) => {
+  return tagState === TagState.CHECKED
+    ? ICONS.CHECKED
+    : tagState === TagState.PARTIAL
+    ? ICONS.PARTIAL
+    : ICONS.UNCHECKED;
+};
+
+const getSelectedAndUnselectedTags = (newOptions: EuiSelectableOption[], tags: State['tags']) => {
+  const selectedTags: string[] = [];
+  const unSelectedTags: string[] = [];
+
+  for (const option of newOptions) {
+    if (option.checked === 'on') {
+      selectedTags.push(option.label);
+    }
+
+    if (!option.checked && tags[option.label].dirty) {
+      unSelectedTags.push(option.label);
+    }
+  }
+
+  return { selectedTags, unSelectedTags };
+};
+
+const NoMatchesMessage: React.FC<{ searchValue: string; onNewItem: (newTag: string) => void }> =
```

Comment:
If the new tag is a substring of an existing tag, there doesn't seem to be a way to add it as new tag. For example - here I want to add a `windows` tag, but as there is a partial match I don't get a link to add it as a new tag:

<img width="454" alt="image" src="https://user-images.githubusercontent.com/7405507/197564329-39f16498-2e07-4b08-88a4-4a0eed2b46b0.png">

Reply:
Great catch! @mdefazio any input on this?

Reply:
Been playing with this, and i'm wondering a few things:
Can we keep the 'Add custom option as tag' close to the input box similar to how the combox box does it?  Clicking the button in the content portion and then having to click to make the focus back on the search box isn't ideal. 
Perhaps to solve the above mentioned issue of substrings, we use a delimiter to mark the creation of the tag? Not sure that's the best—still digging into this and trying a few things.

Reply:
I like the way you have implemented this now with the 'Add xxx as a tag' approach

<img width="353" alt="image" src="https://user-images.githubusercontent.com/7405507/200380572-224a65f7-4b23-4bea-bd07-98191ac7d5cc.png">